File: sdk/kotlin/src/main/kotlin/com/telecom/SecurityAPI.kt
package com.telecom

import kotlinx.serialization.Serializable

/**
 * Security API for fraud detection and SIM swap protection
 */
class SecurityAPI(private val httpClient: HTTPClient) {
    
    /**
     * Analyze a transaction for fraud
     */
    suspend fun analyzeTransaction(transaction: Map<String, Any>): FraudAlert? {
        return httpClient.post("/api/v1/security/fraud/analyze", transaction)
    }
    
    /**
     * Get fraud alerts with filtering
     */
    suspend fun getFraudAlerts(filter: FraudAlertFilter? = null): List<FraudAlert> {
        val payload = filter?.let {
            mapOf(
                "type" to it.type,
                "severity" to it.severity,
                "status" to it.status,
                "limit" to it.limit
            ).filterValues { it != null }
        } ?: emptyMap()
        
        return httpClient.post("/api/v1/security/fraud/alerts", payload)
    }
    
    /**
     * Update fraud alert status
     */
    suspend fun updateAlertStatus(
        alertId: String,
        status: String,
        actions: List<String> = emptyList()
    ): JsonObject {
        return httpClient.put("/api/v1/security/fraud/alerts/$alertId",
            mapOf(
                "status" to status,
                "actions" to actions
            ))
    }
    
    /**
     * Get fraud detection metrics
     */
    suspend fun getFraudMetrics(period: String = "monthly"): FraudMetrics {
        return httpClient.get("/api/v1/security/fraud/metrics",
            mapOf("period" to period))
    }
    
    /**
     * Get detected fraud patterns
     */
    suspend fun getFraudPatterns(): JsonObject {
        return httpClient.get("/api/v1/security/fraud/patterns")
    }
    
    /**
     * Verify SIM swap request
     */
    suspend fun verifySIMSwap(profileId: String, msisdn: String): JsonObject {
        return httpClient.post("/api/v1/security/simswap/verify",
            mapOf(
                "profile_id" to profileId,
                "msisdn" to msisdn
            ))
    }
    
    /**
     * Get SIM swap history for a profile
     */
    suspend fun getSIMSwapHistory(profileId: String): JsonObject {
        return httpClient.get("/api/v1/security/simswap/history/$profileId")
    }
}

@Serializable
data class FraudAlert(
    val id: String,
    val type: String,
    val severity: String,
    val profileId: String,
    val description: String,
    val riskScore: Double,
    val evidence: List<String>,
    val ipAddress: String,
    val timestamp: String,
    val status: String,
    val actionsTaken: List<String>,
    val metadata: Map<String, Any>
)

@Serializable
data class FraudMetrics(
    val period: String,
    val totalAlerts: Long,
    val resolvedAlerts: Long,
    val falsePositives: Long,
    val resolutionRate: Double,
    val falsePositiveRate: Double,
    val byType: Map<String, Long>,
    val bySeverity: Map<String, Long>,
    val generatedAt: String
)

@Serializable
data class FraudAlertFilter(
    val type: String? = null,
    val severity: String? = null,
    val status: String? = null,
    val limit: Int = 50
)

@Serializable
enum class FraudType(val value: String) {
    ACCOUNT_TAKEOVER("account_takeover"),
    SUBSCRIPTION_FRAUD("subscription_fraud"),
    PAYMENT_FRAUD("payment_fraud"),
    USAGE_ANOMALY("usage_anomaly"),
    SIM_SWAP("sim_swap")
}

@Serializable
enum class FraudSeverity(val value: String) {
    LOW("low"),
    MEDIUM("medium"),
    HIGH("high"),
    CRITICAL("critical")
}

File: sdk/swift/Sources/TelecomSDK/SecurityAPI.swift
import Foundation

/// Security API for fraud detection and SIM swap protection
public class SecurityAPI {
    private let client: HTTPClient
    
    public init(client: HTTPClient) {
        self.client = client
    }
    
    /// Analyze a transaction for fraud
    public func analyzeTransaction(_ transaction: [String: Any]) async throws -> FraudAlert? {
        do {
            return try await client.post(
                "/api/v1/security/fraud/analyze",
                body: transaction,
                responseType: FraudAlert.self
            )
        } catch {
            if case HTTPClient.APIError.notFound = error {
                return nil
            }
            throw error
        }
    }
    
    /// Get fraud alerts with filtering
    public func getFraudAlerts(filter: FraudAlertFilter? = nil) async throws -> [FraudAlert] {
        var body: [String: Any] = [:]
        if let filter = filter {
            if let type = filter.type { body["type"] = type }
            if let severity = filter.severity { body["severity"] = severity }
            if let status = filter.status { body["status"] = status }
            body["limit"] = filter.limit
        }
        
        return try await client.post(
            "/api/v1/security/fraud/alerts",
            body: body,
            responseType: [FraudAlert].self
        )
    }
    
    /// Update fraud alert status
    public func updateAlertStatus(
        alertId: String,
        status: String,
        actions: [String] = []
    ) async throws -> [String: Any] {
        return try await client.put(
            "/api/v1/security/fraud/alerts/\(alertId)",
            body: [
                "status": status,
                "actions": actions
            ]
        )
    }
    
    /// Get fraud detection metrics
    public func getFraudMetrics(period: String = "monthly") async throws -> FraudMetrics {
        return try await client.get(
            "/api/v1/security/fraud/metrics",
            parameters: ["period": period],
            responseType: FraudMetrics.self
        )
    }
    
    /// Get detected fraud patterns
    public func getFraudPatterns() async throws -> [String: Any] {
        return try await client.get("/api/v1/security/fraud/patterns")
    }
    
    /// Verify SIM swap request
    public func verifySIMSwap(profileId: String, msisdn: String) async throws -> [String: Any] {
        return try await client.post(
            "/api/v1/security/simswap/verify",
            body: [
                "profile_id": profileId,
                "msisdn": msisdn
            ]
        )
    }
    
    /// Get SIM swap history for a profile
    public func getSIMSwapHistory(profileId: String) async throws -> [String: Any] {
        return try await client.get("/api/v1/security/simswap/history/\(profileId)")
    }
}

public struct FraudAlert: Codable {
    public let id: String
    public let type: String
    public let severity: String
    public let profileId: String
    public let description: String
    public let riskScore: Double
    public let evidence: [String]
    public let ipAddress: String
    public let timestamp: String
    public let status: String
    public let actionsTaken: [String]
    public let metadata: [String: Any]
    
    enum CodingKeys: String, CodingKey {
        case id, type, severity, profileId, description, riskScore
        case evidence, ipAddress, timestamp, status, actionsTaken, metadata
    }
    
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        type = try container.decode(String.self, forKey: .type)
        severity = try container.decode(String.self, forKey: .severity)
        profileId = try container.decode(String.self, forKey: .profileId)
        description = try container.decode(String.self, forKey: .description)
        riskScore = try container.decode(Double.self, forKey: .riskScore)
        evidence = try container.decode([String].self, forKey: .evidence)
        ipAddress = try container.decode(String.self, forKey: .ipAddress)
        timestamp = try container.decode(String.self, forKey: .timestamp)
        status = try container.decode(String.self, forKey: .status)
        actionsTaken = try container.decode([String].self, forKey: .actionsTaken)
        metadata = try container.decode([String: Any].self, forKey: .metadata)
    }
    
    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encode(type, forKey: .type)
        try container.encode(severity, forKey: .severity)
        try container.encode(profileId, forKey: .profileId)
        try container.encode(description, forKey: .description)
        try container.encode(riskScore, forKey: .riskScore)
        try container.encode(evidence, forKey: .evidence)
        try container.encode(ipAddress, forKey: .ipAddress)
        try container.encode(timestamp, forKey: .timestamp)
        try container.encode(status, forKey: .status)
        try container.encode(actionsTaken, forKey: .actionsTaken)
        try container.encode(metadata, forKey: .metadata)
    }
}

public struct FraudMetrics: Codable {
    public let period: String
    public let totalAlerts: Int64
    public let resolvedAlerts: Int64
    public let falsePositives: Int64
    public let resolutionRate: Double
    public let falsePositiveRate: Double
    public let byType: [String: Int64]
    public let bySeverity: [String: Int64]
    public let generatedAt: String
}

public struct FraudAlertFilter {
    public let type: String?
    public let severity: String?
    public let status: String?
    public let limit: Int
    
    public init(type: String? = nil, severity: String? = nil, status: String? = nil, limit: Int = 50) {
        self.type = type
        self.severity = severity
        self.status = status
        self.limit = limit
    }
}

public enum FraudType: String, CaseIterable {
    case accountTakeover = "account_takeover"
    case subscriptionFraud = "subscription_fraud"
    case paymentFraud = "payment_fraud"
    case usageAnomaly = "usage_anomaly"
    case simSwap = "sim_swap"
}

public enum FraudSeverity: String, CaseIterable {
    case low = "low"
    case medium = "medium"
    case high = "high"
    case critical = "critical"
}
